### PR TITLE
Split test frameworks from acceptance test frameworks

### PR DESCRIPTION
**Pull request by carvil**
This PR makes a clearer distinction between testing frameworks (rspec, minitest) and acceptance testing frameworks (capybara/active support test case behaviour). 

It does so by creating acceptance test classes and allow the user to configure which one to use:

```
GovukAbTesting.configure do |config|
  config.acceptance_test_framework = :capybara # or :test_case_behaviour for ActiveSupport request specs
end

# Then in the test file:
include GovukAbTesting::RspecHelpers

# Or
include GovukAbTesting::MinitestHelpers
```

Then, the actual helpers (for either minitest or rspec) will use the corresponding acceptance test framework instance to do things such as set headers, find things on the page, etc.

The reason for this change is that we have a combination of testing frameworks and acceptance testing frameworks in our apps, so this aims to me more flexible.

Trello: https://trello.com/c/t4n8kPpx/434-make-sure-collections-can-be-toggled-with-the-a-b-test-and-react-to-it

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,7 +33,21 @@ To enable testing in the app, your Rails app needs:
...
 3. A response HTTP header that tells Fastly you're doing an A/B test
 
-Let's say you have this controller:
+Start by defining which acceptance testing framework you will use. This gem
+supports both Capybara and ActiveSupport. In order to configure it, add this to
+your test helper file:
+
+```
+GovukAbTesting.configure do |config|
+  config.acceptance_test_framework = :capybara # or :active_support
```
Comment (alecgibson) [inline README.md]: Is there a default? If not, I think there should be. If so, please document which one it is.

File: lib/govuk_ab_testing/acceptance_tests/active_support.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      def set_header(name, value)
+        request.headers[name] = value
+      end
+
+      def vary_header(response)
+        response.headers['Vary']
+      end
+
+      def analytics_meta_tags
+        @analytics_meta_tags ||= scope.css_select("meta[name='govuk:ab-test']")
```
Comment (alecgibson) [inline lib/govuk_ab_testing/acceptance_tests/active_support.rb]: I feel like `meta[name='...']` belongs in a constant somewhere

File: lib/govuk_ab_testing/acceptance_tests/active_support.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      end
+
+      def analytics_meta_tag
+        analytics_meta_tags.first
```
Comment (alecgibson) [inline lib/govuk_ab_testing/acceptance_tests/active_support.rb]: Is `.first` robust to multiple AB tests on the same page?

File: lib/govuk_ab_testing/acceptance_tests/active_support.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      end
+
+      def content
+        analytics_meta_tag.attributes['content'].value
+      end
+
+      def dimension
+        analytics_meta_tag.attributes['data-analytics-dimension'].value
```
Comment (alecgibson) [inline lib/govuk_ab_testing/acceptance_tests/active_support.rb]: `data-analytics-dimension` could also arguably live in a constant

File: lib/govuk_ab_testing/acceptance_tests/active_support.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
+module GovukAbTesting
+  module AcceptanceTests
+    class ActiveSupport
+      attr_reader :request, :scope
+
+      def initialize(scope)
+        @request = scope.instance_variable_get(:@request)
```
Comment (alecgibson) [inline lib/govuk_ab_testing/acceptance_tests/active_support.rb]: Check `@request` is defined/not `nil`? You do a check for `page` in `capybara.rb`

File: lib/govuk_ab_testing/rspec_helpers.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
+module GovukAbTesting
```
Comment (alecgibson) [inline lib/govuk_ab_testing/rspec_helpers.rb]: If we're doing this properly, I think I would personally prefer one more layer of abstraction that would keep the logic synced up between these two test helpers. For example, you might have

```ruby
framework_helper = framework_helper_for(:rspec)

def with_variant(args)
  # Set up variant and AB Test... (omitted)
  framework_helper.set_header(...) # This is implemented by both version of the framework helpers
  
  yield

  framework_helper.assert_vary_header_set(...)
  # etc.
end
```

File: README.md (hunk 1 above)
Comment (carvil) [inline README.md]: I don't think we are in a position to dictate a default because of it varies across apps. The gem will raise an exception if it's not set, which I think is a sensible way of reminding us to set one.

File: lib/govuk_ab_testing/acceptance_tests/active_support.rb (hunk 5 above)
Comment (carvil) [inline lib/govuk_ab_testing/acceptance_tests/active_support.rb]: Good point, will do.

File: lib/govuk_ab_testing/acceptance_tests/active_support.rb (hunk 2 above)
Comment (carvil) [inline lib/govuk_ab_testing/acceptance_tests/active_support.rb]: I can extract it away and make it common between both acceptance test classes.

File: lib/govuk_ab_testing/acceptance_tests/active_support.rb (hunk 3 above)
Comment (carvil) [inline lib/govuk_ab_testing/acceptance_tests/active_support.rb]: Probably now, but we assert there is only one in both test helpers, which is our use case now. I think we shouldn't make it work with more A/B tests for now until we have that need if that makes sense.

File: lib/govuk_ab_testing/acceptance_tests/active_support.rb (hunk 4 above)
Comment (carvil) [inline lib/govuk_ab_testing/acceptance_tests/active_support.rb]: Ok, I will make it a part of the other one 👍

File: README.md (hunk 1 above)
Comment (alecgibson) [inline README.md]: Sure fine to leave.

File: lib/govuk_ab_testing/acceptance_tests/active_support.rb (hunk 3 above)
Comment (alecgibson) [inline lib/govuk_ab_testing/acceptance_tests/active_support.rb]: Sure. Can we document the fact that it'll only work with one AB test in the `README`?

Comment (carvil) [inline lib/govuk_ab_testing/acceptance_tests/active_support.rb]: Yup, will do.

File: lib/govuk_ab_testing/rspec_helpers.rb (hunk 6 above)
Comment (carvil) [inline lib/govuk_ab_testing/rspec_helpers.rb]: I can see how we could remove some duplication by introducing this change. However, I would rather do itwhen we introduce another test helper class, so we have a 3rd example on how to implement `with_variant`. I don't know enough about its usage at this stage to know if that would be a good way forward or not.

Comment (alecgibson) [inline lib/govuk_ab_testing/rspec_helpers.rb]: Yes possibly - I thought about this a bit more and maybe you could move all the setup and assertion steps into `active_support.rb` and `capybara.rb`? Then you could just have a single helper file (or two very lightweight helpers that point to a common one with different initialization). That would avoid the need for another layer of abstraction, and just slightly move the boundary of abstraction in a way that will keep the testing logic consistent.

Comment (carvil) [inline lib/govuk_ab_testing/rspec_helpers.rb]: As discussed IRL, we will do this refactoring in a separate PR.

File: lib/govuk_ab_testing/requested_variant.rb [hunk 7]
```diff
@@ -1,23 +1,23 @@
 module GovukAbTesting
   class RequestedVariant
-    attr_reader :ab_test, :request
+    attr_reader :ab_test, :request_headers
```
Comment (alecgibson) [inline lib/govuk_ab_testing/requested_variant.rb]: This is a breaking change, so please make sure you issue a major version bump.

Comment (carvil) [inline lib/govuk_ab_testing/requested_variant.rb]: Will do, thanks!
